src/main/java/spoon/reflect/visitor/ImportScannerImpl.java

```diff
@@ -218,7 +226,44 @@ private boolean addImport(CtTypeReference<?> ref) {
 		if (imports.containsKey(ref.getSimpleName())) {
 			return isImported(ref);
 		}
+		// don't import unnamed package elements
+		if (ref.getPackage() == null || ref.getPackage().isUnnamedPackage()) {
+			return false;
+		}
+		if (!ref.getPackage().getSimpleName().equals("java.lang")) {
+			if (classNamePresentInJavaLang(ref)) {
+				// Don't import class with names clashing with some classes present in java.lang,
+				// because it leads to undecidability and compilation errors. I. e. always leave
+				// com.mycompany.String fully-qualified.
+				return false;
+			}
+		}
 		imports.put(ref.getSimpleName(), ref);
 		return true;
 	}
+
+	private boolean classNamePresentInJavaLang(CtTypeReference<?> ref) {
+		Boolean presentInJavaLang = namesPresentInJavaLang.get(ref.getSimpleName());
+		if (presentInJavaLang == null) {
+			// The following procedure of determining if the handle is present in Java Lang or
+			// not produces "false positives" if the analyzed source complianceLevel is > 6.
+			// For example, it reports that FunctionalInterface is present in java.lang even
+			// for compliance levels 6, 7. But this is not considered a bad thing, in opposite,
+			// it makes generated code a little more compatible with future versions of Java.
+			if (namesPresentInJavaLang8.contains(ref.getSimpleName())
+					|| namesPresentInJavaLang9.contains(ref.getSimpleName())) {
+				presentInJavaLang = true;
+			} else {
+				// Assuming Spoon's own runtime environment is Java 7+
+				try {
```

Comment:
If spoon's runtime is Java 7 and analyzed code is Java 8 it won't skip e. g. `FunctionalInterface`.